File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/547bd2425371d0161e1e62d1a00e6d15.py
import sys
from collections import defaultdict
from fractions import Fraction

mod = 998244353

class ModInt:
    def __init__(self, x=0):
        self.x = x % mod

    def __add__(self, other):
        return ModInt(self.x + other.x)

    def __sub__(self, other):
        return ModInt(self.x - other.x)

    def __mul__(self, other):
        return ModInt(self.x * other.x)

    def __truediv__(self, other):
        return self * other.inverse()

    def inverse(self):
        return self.pow(mod - 2)

    def pow(self, n):
        result = ModInt(1)
        base = self
        while n > 0:
            if n & 1:
                result *= base
            base *= base
            n >>= 1
        return result

    def __repr__(self):
        return str(self.x)

class Matrix:
    def __init__(self):
        self.A = [[ModInt() for _ in range(4)] for _ in range(4)]

    @staticmethod
    def I():
        mat = Matrix()
        for i in range(4):
            mat.A[i][i] = ModInt(1)
        return mat

    def __mul__(self, other):
        C = Matrix()
        for i in range(4):
            for j in range(4):
                for k in range(4):
                    C.A[i][j] += self.A[i][k] * other.A[k][j]
        return C

    def __getitem__(self, idx):
        return self.A[idx]

class SegmentTree:
    def __init__(self, n, f, M1):
        self.sz = 1
        while self.sz < n:
            self.sz <<= 1
        self.seg = [M1 for _ in range(2 * self.sz)]
        self.f = f
        self.M1 = M1

    def set(self, k, x):
        self.seg[k + self.sz] = x

    def build(self):
        for k in range(self.sz - 1, 0, -1):
            self.seg[k] = self.f(self.seg[2 * k], self.seg[2 * k + 1])

    def update(self, k, x):
        k += self.sz
        self.seg[k] = x
        while k > 1:
            k >>= 1
            self.seg[k] = self.f(self.seg[2 * k], self.seg[2 * k + 1])

    def query(self, a, b):
        L = self.M1
        R = self.M1
        a += self.sz
        b += self.sz
        while a < b:
            if a & 1:
                L = self.f(L, self.seg[a])
                a += 1
            if b & 1:
                b -= 1
                R = self.f(self.seg[b], R)
            a >>= 1
            b >>= 1
        return self.f(L, R)

def main():
    N = int(sys.stdin.readline().strip())
    X = []
    V = []
    P = []
    Q = []
    
    for _ in range(N):
        x, v, p = map(int, sys.stdin.readline().strip().split())
        X.append(x)
        V.append(v)
        P.append(ModInt(p) / ModInt(100))
        Q.append(ModInt(1) - P[-1])

    ev = defaultdict(list)
    for i in range(1, N):
        if V[i - 1] < V[i]:
            ev[Fraction(X[i - 1] - X[i], V[i - 1] - V[i])].append((i, 0))
        ev[Fraction(X[i] - X[i - 1], V[i - 1] + V[i])].append((i, 2))
        if V[i - 1] > V[i]:
            ev[Fraction(X[i - 1] - X[i], V[i] - V[i - 1])].append((i, 1))

    Mat = Matrix
    e = Mat.I()
    f = lambda a, b: a * b

    seg = SegmentTree(N, f, e)
    for i in range(N):
        u = Mat()
        u[0][0] = Q[i]
        u[1][0] = Q[i]
        u[2][2] = Q[i]
        u[3][2] = Q[i]
        u[0][1] = P[i]
        u[1][1] = P[i]
        u[2][3] = P[i]
        u[3][3] = P[i]
        seg.set(i, u)
    seg.build()

    ret = ModInt(0)

    for p, vs in ev.items():
        for tt in vs:
            now = seg.seg[tt[0]]
            if tt[1] == 0:
                now[0][0] = ModInt(0)
                now[0][2] = Q[tt[0]]
            elif tt[1] == 1:
                now[1][1] = ModInt(0)
                now[1][3] = P[tt[0]]
            else:
                now[1][0] = ModInt(0)
                now[1][2] = Q[tt[0]]
            seg.update(tt[0], now)

        uku = seg.seg[1]
        sum_ = uku[0][2] + uku[0][3]
        ret += ModInt(p.numerator) / ModInt(p.denominator) * sum_

        for tt in vs:
            now = seg.seg[tt[0]]
            if tt[1] == 0:
                now[0][2] = ModInt(0)
                now[2][2] = ModInt(0)
            elif tt[1] == 1:
                now[1][3] = ModInt(0)
                now[3][3] = ModInt(0)
            else:
                now[1][2] = ModInt(0)
                now[3][2] = ModInt(0)
            seg.update(tt[0], now)

    print(ret)

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/547bd2425371d0161e1e62d1a00e6d15.go
package main

import (
	"fmt"
	"math/big"
	"sort"
)

const mod = 998244353

type ModInt struct {
	x uint32
}

func NewModInt(val int64) ModInt {
	return ModInt{reduce(uint64(val) * r2)}
}

func (m ModInt) Get() int {
	return int(reduce(uint64(m.x))) % mod
}

func (m ModInt) Add(other ModInt) ModInt {
	m.x += other.x
	if int(m.x) >= 2*mod {
		m.x -= 2 * mod
	}
	return m
}

func (m ModInt) Sub(other ModInt) ModInt {
	m.x -= other.x
	if int(m.x) < 0 {
		m.x += 2 * mod
	}
	return m
}

func (m ModInt) Mul(other ModInt) ModInt {
	m.x = reduce(uint64(m.x) * uint64(other.x))
	return m
}

func (m ModInt) Pow(n int64) ModInt {
	ret := NewModInt(1)
	mul := m
	for n > 0 {
		if n&1 == 1 {
			ret = ret.Mul(mul)
		}
		mul = mul.Mul(mul)
		n >>= 1
	}
	return ret
}

func (m ModInt) Inverse() ModInt {
	return m.Pow(mod - 2)
}

func reduce(w uint64) uint32 {
	return uint32(w >> 32) + mod - uint32((uint64(uint32(w)*inv)*mod)>>32)
}

var inv uint32
var r2 uint32

func init() {
	inv = mulInv(mod)
	r2 = uint32(-int64(mod) % mod)
}

func mulInv(n uint32) uint32 {
	x := uint32(1)
	for e := 5; e > 0; e-- {
		x = x * (2 - x*n)
	}
	return x
}

type Fraction struct {
	num, dom int64
}

func NewFraction(n, d int64) Fraction {
	if d == 0 {
		panic("denominator cannot be zero")
	}
	if d < 0 {
		n *= -1
		d *= -1
	}
	gcd := gcd(abs(n), abs(d))
	return Fraction{n / gcd, d / gcd}
}

func gcd(a, b int64) int64 {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func (f Fraction) Add(other Fraction) Fraction {
	return NewFraction(f.num*other.dom+other.num*f.dom, f.dom*other.dom)
}

func (f Fraction) Sub(other Fraction) Fraction {
	return NewFraction(f.num*other.dom-other.num*f.dom, f.dom*other.dom)
}

func (f Fraction) Mul(other Fraction) Fraction {
	return NewFraction(f.num*other.num, f.dom*other.dom)
}

func (f Fraction) Div(other Fraction) Fraction {
	return NewFraction(f.num*other.dom, f.dom*other.num)
}

type Node struct {
	right int
	typ   int
}

type Event struct {
	fraction Fraction
	nodes    []Node
}

type SegmentTree struct {
	sz  int
	seg []Matrix
	f   func(a, b Matrix) Matrix
	m1  Matrix
}

func NewSegmentTree(n int, f func(a, b Matrix) Matrix, m1 Matrix) SegmentTree {
	sz := 1
	for sz < n {
		sz <<= 1
	}
	seg := make([]Matrix, 2*sz)
	for i := range seg {
		seg[i] = m1
	}
	return SegmentTree{sz: sz, seg: seg, f: f, m1: m1}
}

func (st *SegmentTree) Set(k int, x Matrix) {
	st.seg[k+st.sz] = x
}

func (st *SegmentTree) Build() {
	for k := st.sz - 1; k > 0; k-- {
		st.seg[k] = st.f(st.seg[2*k], st.seg[2*k+1])
	}
}

func (st *SegmentTree) Update(k int, x Matrix) {
	k += st.sz
	st.seg[k] = x
	for k >>= 1; k > 0; k >>= 1 {
		st.seg[k] = st.f(st.seg[2*k], st.seg[2*k+1])
	}
}

func (st *SegmentTree) Query(a, b int) Matrix {
	L, R := st.m1, st.m1
	for a += st.sz; a < b; a >>= 1 {
		if a&1 != 0 {
			L = st.f(L, st.seg[a])
			a++
		}
	}
	for b += st.sz; b > 0; b >>= 1 {
		if b&1 != 0 {
			b--
			R = st.f(st.seg[b], R)
		}
	}
	return st.f(L, R)
}

type Matrix struct {
	A [4][4]ModInt
}

func NewMatrix() Matrix {
	return Matrix{}
}

func (m *Matrix) Identity() {
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			if i == j {
				m.A[i][j] = NewModInt(1)
			} else {
				m.A[i][j] = NewModInt(0)
			}
		}
	}
}

func (m *Matrix) Mul(other Matrix) Matrix {
	var C Matrix
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			for k := 0; k < 4; k++ {
				C.A[i][j] = C.A[i][j].Add(m.A[i][k].Mul(other.A[k][j]))
			}
		}
	}
	return C
}

func (m *Matrix) Pow(k int64) Matrix {
	B := NewMatrix()
	B.Identity()
	for k > 0 {
		if k&1 == 1 {
			B = B.Mul(*m)
		}
		*m = m.Mul(*m)
		k >>= 1
	}
	return B
}

func main() {
	var N int
	fmt.Scan(&N)
	X := make([]int, N)
	V := make([]int, N)
	P := make([]ModInt, N)
	Q := make([]ModInt, N)

	for i := 0; i < N; i++ {
		fmt.Scan(&X[i], &V[i])
		var p int64
		fmt.Scan(&p)
		P[i] = NewModInt(p).Div(NewModInt(100))
		Q[i] = NewModInt(1).Sub(P[i])
	}

	ev := make(map[Fraction][]Node)

	for i := 1; i < N; i++ {
		if V[i-1] < V[i] {
			ev[NewFraction(int64(X[i-1]-X[i]), int64(V[i-1]-V[i]))] = append(ev[NewFraction(int64(X[i-1]-X[i]), int64(V[i-1]-V[i]))], Node{i, 0})
		}
		ev[NewFraction(int64(X[i]-X[i-1]), int64(V[i-1]+V[i]))] = append(ev[NewFraction(int64(X[i]-X[i-1]), int64(V[i-1]+V[i]))], Node{i, 2})
		if V[i-1] > V[i] {
			ev[NewFraction(int64(X[i-1]-X[i]), int64(V[i]-V[i-1]))] = append(ev[NewFraction(int64(X[i-1]-X[i]), int64(V[i]-V[i-1]))], Node{i, 1})
		}
	}

	mat := NewMatrix()
	mat.Identity()
	seg := NewSegmentTree(N, func(a, b Matrix) Matrix {
		return a.Mul(b)
	}, mat)

	for i := 0; i < N; i++ {
		u := NewMatrix()
		u.Identity()
		u.A[0][0] = Q[i]
		u.A[1][0] = Q[i]
		u.A[2][2] = Q[i]
		u.A[3][2] = Q[i]
		u.A[0][1] = P[i]
		u.A[1][1] = P[i]
		u.A[2][3] = P[i]
		u.A[3][3] = P[i]
		seg.Set(i, u)
	}
	seg.Build()

	ret := NewModInt(0)

	for _, p := range ev {
		for _, tt := range p {
			now := seg.seg[tt.right]
			if tt.typ == 0 {
				now.A[0][0] = NewModInt(0)
				now.A[0][2] = Q[tt.right]
			} else if tt.typ == 1 {
				now.A[1][1] = NewModInt(0)
				now.A[1][3] = P[tt.right]
			} else {
				now.A[1][0] = NewModInt(0)
				now.A[1][2] = Q[tt.right]
			}
			seg.Update(tt.right, now)
		}
		uku := seg.seg[1]
		sum := uku.A[0][2].Add(uku.A[0][3])
		ret = ret.Add(NewModInt(p.fraction.num).Div(p.fraction.dom).Mul(sum))
		for _, tt := range p {
			now := seg.seg[tt.right]
			if tt.typ == 0 {
				now.A[0][2] = NewModInt(0)
				now.A[2][2] = NewModInt(0)
			} else if tt.typ == 1 {
				now.A[1][3] = NewModInt(0)
				now.A[3][3] = NewModInt(0)
			} else {
				now.A[1][2] = NewModInt(0)
				now.A[3][2] = NewModInt(0)
			}
			seg.Update(tt.right, now)
		}
	}
	fmt.Println(ret.Get())
}
